### PR TITLE
[gitlab] Send pipeline detailed data to Datadog

### PR DESCRIPTION
**Pull request by KSerrania**
### What does this PR do?

This PR does two things:
- it refactors the notification tasks: reusable functions have been refactored in a new `tasks/libs/pipeline_data.py` file, the notification Gitlab jobs have been merged into one job that handles all cases.
- it creates a new task (and associated job) that gathers statistics about the reasons why a pipeline failed, expanding on the work done in #13005, and sends them to Datadog, on all pipelines. This uses the official Datadog Python API client.

### Motivation

Get more detailed statistics on why pipelines fail. In particular, get statistics on infrastructure errors, which can only be discriminated by parsing the job logs.

Example dashboard with the data from this PR:
image
image

### Possible Drawbacks / Trade-offs

n/a

### Describe how to test/QA your changes

Once this is merged, check that we do receive data from pipelines.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: .gitlab/notify.yml [hunk 1]
```diff
@@ -54,10 +31,31 @@ notify-on-failure:
...
+  when: always
+  dependencies: []
+  script:
+    - source /root/.bashrc && conda activate ddpy3
+    - set +x
+    - export GITLAB_TOKEN=$(aws ssm get-parameter --region us-east-1 --name ci.datadog-agent.gitlab_read_api_token --with-decryption --query "Parameter.Value" --out text)
+    - export DD_API_KEY=$(aws ssm get-parameter --region us-east-1 --name ci.datadog-agent.datadog_api_key_org2 --with-decryption --query "Parameter.Value" --out text)
+    - python3 -m pip install -r tasks/libs/requirements-stats.txt
+    - |
+      # Do not send statistics if this is a child pipeline of another repo
```
Comment (bkabrda) [inline .gitlab/notify.yml:56]: If the aim is to get statistics on failures, why does it matter if this was triggered as a child pipeline of another repo?

File: tasks/libs/pipeline_data.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+def get_job_failure_context(job_log):
+    """
+    Parses job logs (provided as a string), and returns the type of failure (infra or job) as well
+    as the precise reason why the job failed.
+    """
+
+    infra_failure_logs = [
+        # Gitlab errors while pulling image on legacy runners
+        ("no basic auth credentials (manager.go:203:0s)", FailedJobReason.MAIN_RUNNER),
+        ("net/http: TLS handshake timeout (manager.go:203:10s)", FailedJobReason.MAIN_RUNNER),
```
Comment (bkabrda) [inline tasks/libs/pipeline_data.py:63]: How stable are these `manager.go:203:...` lines? Would it make more sense to make these regex that would match anything in the parentheses so we don't miss the error if the `manager.go` file changes or the code moves to a different file?

File: tasks/pipeline.py [hunk 3]
```diff
@@ -437,47 +434,146 @@ def trigger_child_pipeline(_, git_ref, project_name, variables="", follow=True):
...
+
+    try:
+        job_failure_stats = get_failed_jobs_stats(project_name, os.getenv("CI_PIPELINE_ID"))
+    except Exception as e:
+        print("Found exception when generating statistics:")
+        print(e)
+        traceback.print_exc(limit=-1)
+        raise Exit(code=1)
+
+    if not os.environ.get("DD_API_KEY"):
```
Comment (bkabrda) [inline tasks/pipeline.py:511]:
```suggestion
    if not (print_to_stdout or os.environ.get("DD_API_KEY")):
```

**Review by bkabrda — DISMISSED** [review]
Nice work. I left couple minor notes inline to address, but mostly this looks really good!

File: tasks/libs/pipeline_data.py (hunk 2 above)
Comment (KSerrania) [reply tasks/libs/pipeline_data.py]: They're stable as long as we don't change Gitlab versions, so I made them more generic.

File: .gitlab/notify.yml (hunk 1 above)
Comment (KSerrania) [reply .gitlab/notify.yml:56]: That was a copy-paste from the notification jobs, I'll remove that.

File: tasks/libs/datadog_api.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+def create_count(metric_name, timestamp, value, tags):
```
Comment (mx-psi) [inline tasks/libs/datadog_api.py:1]: Should we calculate the timestamp inside this function?

File: tasks/libs/pipeline_stats.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    #   }
+    # }
+    job_failure_stats = defaultdict(dict)
+
+    failed_jobs = get_failed_jobs(project_name, pipeline_id)
+
+    for job in failed_jobs:
+        job_failure_stats[job["failure_type"]][job["failure_reason"]] = (
+            job_failure_stats[job["failure_type"]].get(job["failure_reason"], 0) + 1
+        )
```
Comment (mx-psi) [inline tasks/libs/pipeline_stats.py:27]: You can use a `Counter` from `collections` to simplify this
```suggestion
    job_failure_stats = defaultdict(Counter)

    failed_jobs = get_failed_jobs(project_name, pipeline_id)

    for job in failed_jobs:
        job_failure_stats[job["failure_type"]][job["failure_reason"]] += 1
```

File: tasks/libs/types.py [hunk 6]
```diff
@@ -38,22 +38,30 @@ def key(self):
...
     INFRA_FAILURES_SECTION_HEADER = "Infrastructure failures:"
     TEST_SECTION_HEADER = "Failed unit tests:"
     MAX_JOBS_PER_TEST = 2
 
-    def __init__(self, base_message, jobs=None):
-        self.base_message = base_message
-        self.failed_jobs = [job for job in jobs if job["failure_type"] == FailedJobReason.JOB_FAILURE] if jobs else []
+    def __init__(self, base="", jobs=None):
+        self.base_message = base
+        self.failed_jobs = [job for job in jobs if job["failure_type"] == FailedJobType.JOB_FAILURE] if jobs else []
```
Comment (mx-psi) [inline tasks/libs/types.py:62]: nit: maybe this to simplify the two assignments below?
```suggestion
    def __init__(self, base="", jobs=None):
        jobs = jobs if jobs else []
        self.base_message = base
        self.failed_jobs = [job for job in jobs if job["failure_type"] == FailedJobType.JOB_FAILURE]
```

File: tasks/pipeline.py [hunk 7]
```diff
@@ -346,13 +344,12 @@ def wait_for_pipeline_from_ref(gitlab, ref):
...
-def generate_failure_messages(base):
-    project_name = "DataDog/datadog-agent"
+def generate_failure_messages(project_name, failed_jobs):
     all_teams = "@DataDog/agent-all"
-    failed_jobs = get_failed_jobs(project_name, os.getenv("CI_PIPELINE_ID"))
+
     # Generate messages for each team
-    messages_to_send = defaultdict(lambda: TeamMessage(base))
-    messages_to_send[all_teams] = SlackMessage(base, jobs=failed_jobs)
+    messages_to_send = defaultdict(lambda: TeamMessage())
```
Comment (mx-psi) [inline tasks/pipeline.py:351]:
```suggestion
    messages_to_send = defaultdict(TeamMessage)
```

File: tasks/libs/datadog_api.py (hunk 4 above)
Comment (KSerrania) [reply tasks/libs/datadog_api.py:1]: I want to ensure that all the metrics I'm creating in a run of the task have the same timestamp, which is why I opted for creating the timestamp beforehand and passing it here.

Comment (mx-psi) [reply tasks/libs/datadog_api.py:1]: Okay, that makes sense

File: tasks/libs/types.py [hunk 8]
```diff
@@ -38,22 +38,31 @@ def key(self):
...
-    def __init__(self, base_message, jobs=None):
-        self.base_message = base_message
-        self.failed_jobs = [job for job in jobs if job["failure_type"] == FailedJobReason.JOB_FAILURE] if jobs else []
+    def __init__(self, base="", jobs=None):
+        jobs = jobs if jobs else []
+        self.base_message = base
+        self.failed_jobs = [job for job in jobs if job["failure_type"] == FailedJobType.JOB_FAILURE]
         self.infra_failed_jobs = (
-            [job for job in jobs if job["failure_type"] == FailedJobReason.INFRA_FAILURE] if jobs else []
+            [job for job in jobs if job["failure_type"] == FailedJobType.INFRA_FAILURE] if jobs else []
```
Comment (mx-psi) [inline tasks/libs/types.py:65]:
```suggestion
            [job for job in jobs if job["failure_type"] == FailedJobType.INFRA_FAILURE]
```

**Review by mx-psi — APPROVED** [review]

**Pull request merged**